Studio/src/Application/Data/MeshWarper.cpp

```diff
@@ -216,6 +303,64 @@ Eigen::MatrixXd MeshWarper::remove_bad_particles(const Eigen::MatrixXd& particle
   return new_particles;
 }
 
+//---------------------------------------------------------------------------
+void MeshWarper::split_cell_on_edge(int cell_id, int new_vertex, int v0, int v1)
+{
+  vtkCell* cell = this->reference_mesh_->GetCell(cell_id);
+  int p0 = cell->GetPointId(0);
+  int p1 = cell->GetPointId(1);
+  int p2 = cell->GetPointId(2);
+  int edge = -1;
+  if ((v0 == p0 && v1 == p1) || (v0 == p1 && v1 == p0)) {
+    edge = 0;
+  }
+  else if ((v0 == p1 && v1 == p2) || (v0 == p2 && v1 == p1)) {
+    edge = 1;
+  }
+  else if ((v0 == p0 && v1 == p2) || (v0 == p2 && v1 == p0)) {
+    edge = 2;
+  }
+
+  vtkSmartPointer<vtkIdList> list = vtkSmartPointer<vtkIdList>::New();
+  list->SetNumberOfIds(3);
+  if (edge == 0) {
+    list->SetId(0, new_vertex);
+    list->SetId(1, cell->GetPointId(1));
+    list->SetId(2, cell->GetPointId(2));
+    this->reference_mesh_->InsertNextCell(VTK_TRIANGLE, list);
+
+    list->SetId(0, new_vertex);
+    list->SetId(1, cell->GetPointId(2));
+    list->SetId(2, cell->GetPointId(0));
+    this->reference_mesh_->InsertNextCell(VTK_TRIANGLE, list);
+  }
+  else if (edge == 1) {
+    list->SetId(0, new_vertex);
+    list->SetId(1, cell->GetPointId(0));
+    list->SetId(2, cell->GetPointId(1));
```

Comment:
should it be GetPointId(2)?

(also, is it possible to use modulo arithmetic in this section? maybe something like GetPointId((edge+1)%3) and GetPointId((edge+2)%3) but with the correct ordering)

Reply:
There are 3 cases:

<img width="518" alt="image" src="https://user-images.githubusercontent.com/1693349/116193981-6e204280-a6ed-11eb-8ae9-fc6d0e4a407e.png">

It can definitely be simplified since two triangles are repeated in each case.  It can be change to a **not** and only one triangle per section.

Are you sure that it's wrong?

Reply:
Oh I see,  I  messed up the edge conventions. I thought edge==1 implies that the edge is opposite to vertex 1

Reply:
Is that the preferred convention?  If so, we should probably change to conform.

Reply:
I don't know if there is a preferred convention